datastore/adapter/keytransform.py

```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
+
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to initialize an `Adapter`
+		*args
+			Variable length list argument to be passed to cls constructor
+		sharding_key
+			Key used to access the sharding function in the underlying Datastore
+		default_sharding_func
+			Default sharding func to use if one is not found in the datastore
+		"""
+		# Get the variables needed to parse the sharding function
+		sharding_key = (
+			sharding_key
+			if sharding_key is not None
+			else cls._default_sharding_key  # type: ignore[attr-defined] # noqa: F821
+		)
+		default_sharding_func = (
+			default_sharding_func
+			if default_sharding_func is not None
+			else f"{cls.prefix}{cls._default_sharding_func}"   # type: ignore[attr-defined] # noqa: F821
+		)
+		sharding_fn: KEY_TRANSFORM_T = await \
+			cls._parse_sharding_function(  # type: ignore[attr-defined] # noqa: F821
+				child_datastore, sharding_key, default_sharding_func
+			)
+
+		# Return an instantiated version of the class
+		return cls(
+			child_datastore, *args, key_transform=sharding_fn, **kwargs
+			)  # type: ignore[attr-defined] # noqa: F821
+
+	@staticmethod
+	def _prefix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).ljust(length, "_")[:length])
+	
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	@classmethod
+	async def _parse_sharding_function(
+			cls: typing.Type[T],
+			child_datastore: DS,
+			sharding_key: datastore.Key,
+			default_sharding_func: str
+	) -> KEY_TRANSFORM_T:
+		"""Determine the sharding function from the sharding_key.
+		
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to access the `sharding_key`
+		sharding_key
+			Key used to access the sharding function in the `child_datastore`
+		default_sharding_func
+			Default sharding func to use if one is not found in `child_datastore`
+
+		Raises
+		------
+		Exception
+			Empty shard identifier file.
+		Exception
+			Prefix was not present in sharding file
+		Exception
+			Sharding file was in incorrect format
+		Exception
+			Expect `v1` format
+		Exception
+			`key_length` shouold be an integer
+		Exception
+			`function_name` was not `prefix`, `suffix`, or `next-to-last`
+		RuntimeError
+			An internal error occurred
+		"""
+
+		prefix: str = cls.prefix  # type: ignore[attr-defined] # noqa: F821
+
+		sharding_func: str
+		need_to_store_sharding_func: bool = False
+		try:
+			# Try reading existing sharding func used
+			sharding_func = next(   # type: ignore[assignment] # noqa: F821
+				iter(await child_datastore.get_all(sharding_key))
+			)
```

Comment:
This looks like it will return a binary string / bytes object when `self` is `BinaryAdapter`, but you need a Unicode string so you'll need to call `decode` somewhere. I propose the following to support for object and binary datastores:

```py
if isinstance(self, datastore.abc.BinaryAdapter):
	# Decode using latin-1 to get a 1:1 binary to Unicode mapping that cannot fail
	sharding_func = (await child_datastore.get_all(sharding_key)).decode("latin-1")
else:
	# Expect result from object datastore be a Unicode string
	sharding_func = typing.cast(str, await child_datastore.get_all(sharding_key))
	assert isinstance(sharding_func, str)
```